File: frontend/src/api/examples.ts
/**
 * The built-in example catalogue, which lives on the backend.
 *
 * Examples are authored as OpenQASM and parsed by the same importer a user's
 * file goes through -- `Roadmap.md`'s exit criterion, and the reason this is a
 * network call rather than a bundled JSON blob. Shipping the circuits in the
 * frontend would make them the only circuits in the app that never passed
 * through the import path, which is exactly the evidence they exist to be.
 *
 * **Two calls, not one.** The catalogue carries metadata and a circuit is
 * fetched when one is chosen. Six small circuits would fit in a single
 * response; [ADR-0009](../../../docs/decisions/ADR0009_CircuitCatalogue.md)
 * section 3 keeps them apart because a *generated* entry -- a QAOA layer count,
 * a VQE ansatz width -- has no single circuit to bundle, and the list's contract
 * should not have to change when one arrives.
 *
 * Throws `ApiError` like every other module here. The editor converts it, which
 * keeps transport-failure-to-user-message in one place.
 */

import type { Circuit } from '../model/circuit';
import { ApiError, request } from './client';

const USE_MOCK = import.meta.env.VITE_USE_MOCK_API === 'true';

/**
 * One catalogue entry.
 *
 * `parameters` is deliberately absent rather than declared and empty. ADR-0009
 * section 4 leaves room for generated entries to carry one, and a client that
 * ignores a field it does not know about keeps working when they do.
 */
export interface ExampleEntry {
  readonly id: string;
  readonly name: string;
  readonly summary: string;
  readonly qubitCount: number;
  readonly operationCount: number;
}

interface CatalogueResponse {
  readonly examples: readonly ExampleEntry[];
}

interface CircuitResponse {
  readonly circuit: Circuit;
}

function unavailable(): ApiError {
  /**
   * No mock, for the reason OpenQASM import has none: a mock would need a copy
   * of the catalogue in the frontend, which is the duplication that keeping it
   * on the backend exists to avoid.
   */
  return new ApiError(
    'BACKEND_UNAVAILABLE',
    'Example circuits need the backend, and the app is running against mocks.',
    0,
  );
}

export async function fetchExamples(
  signal?: AbortSignal,
): Promise<readonly ExampleEntry[]> {
  if (USE_MOCK) throw unavailable();

  const response = await request<CatalogueResponse>('/examples', {
    ...(signal ? { signal } : {}),
  });

  return response.examples;
}

export async function fetchExample(
  id: string,
  signal?: AbortSignal,
): Promise<Circuit> {
  if (USE_MOCK) throw unavailable();

  const response = await request<CircuitResponse>(
    `/examples/${encodeURIComponent(id)}`,
    { ...(signal ? { signal } : {}) },
  );

  return response.circuit;
}

File: backend/src/phasor_workbench/api/routes/examples.py
"""The built-in example catalogue.

Two endpoints rather than one, per
[ADR-0009](../../../../../docs/decisions/ADR0009_CircuitCatalogue.md): the list
carries metadata and the circuit is fetched when something is chosen. Six small
circuits would fit in one response, and the split is not about size -- it is
what lets a generated entry, which has no single circuit, join the same list
without changing its contract.

Circuits go through `read_circuit`, the same call an uploaded file makes. An
example that stopped parsing would raise rather than 404, because that is a
broken build rather than a missing resource, and it is what the suite asserts
against for every example.
"""

from __future__ import annotations

from typing import Any

from fastapi import APIRouter
from pydantic import BaseModel, ConfigDict, Field

from ...examples import catalogue, find
from ..documents import read_circuit
from ..errors import ApiError, ErrorCode

router = APIRouter(tags=["examples"])


class ExampleEntry(BaseModel):
    """One catalogue entry.

    No circuit, and no `parameters` yet. ADR-0009 section 4 keeps room for the
    second: a client that ignores an absent field keeps working when generated
    entries start carrying one.
    """

    model_config = ConfigDict(populate_by_name=True)

    id: str
    name: str
    summary: str
    qubit_count: int = Field(serialization_alias="qubitCount")
    operation_count: int = Field(serialization_alias="operationCount")


class ExampleCatalogueResponse(BaseModel):
    examples: list[ExampleEntry]


class ExampleCircuitResponse(BaseModel):
    circuit: dict[str, Any]


@router.get(
    "/examples",
    response_model=ExampleCatalogueResponse,
    response_model_by_alias=True,
    summary="Built-in example circuits",
)
def get_examples() -> ExampleCatalogueResponse:
    """List the catalogue, ordered by id so a picker is stable between loads.

    The counts are computed rather than declared in the files, so they cannot
    disagree with the circuit -- the same reason nothing else here is written
    down twice.
    """
    entries = []

    for example in catalogue():
        circuit = read_circuit(example.document())
        entries.append(
            ExampleEntry(
                id=example.id,
                name=example.name,
                summary=example.summary,
                qubit_count=len(circuit.qubits),
                operation_count=len(circuit.operations),
            )
        )

    return ExampleCatalogueResponse(examples=entries)


@router.get(
    "/examples/{identifier}",
    response_model=ExampleCircuitResponse,
    summary="One example as a circuit document",
    responses={404: {"description": "No example with that id."}},
)
def get_example(identifier: str) -> ExampleCircuitResponse:
    """One example, in the wire form every other endpoint accepts."""
    example = find(identifier)

    if example is None:
        raise ApiError(
            code=ErrorCode.NOT_FOUND,
            message=f"No example circuit with id '{identifier}'.",
            status_code=404,
        )

    circuit = read_circuit(example.document())

    return ExampleCircuitResponse(
        circuit=circuit.model_dump(by_alias=True, mode="json")
    )
